### ISSUE TITLE
[UBSan] Regression: Bogus "insufficient space for object" error with struct containing 0-length array

### ISSUE THREAD

**Issue by BertalanD** [llvm/llvm-project#61775]
*Reproducer on Godbolt: https://godbolt.org/z/sfjrvnM7f*

Consider the following code:

```c++
#include <span>

template<typename T, size_t N>
struct my_array {
    T m_data[N];
    operator std::span<T>() { return { m_data, N }; }
};

struct base {
    std::span<int> m_values;
};

template<typename... Args>
struct storage : base {
    storage(Args const&... args)
        : m_data({ args... })
    {
        m_values = m_data;
    }
    my_array<int, sizeof...(Args)> m_data;
};

int main() {
    storage<> bug_struct;
}
```

When compiling this with Clang 16 with the UBSan object size checker enabled:
```console
clang++-16 -std=c++20 -fsanitize=object-size
```
The following error is produced at runtime:
```
/app/example.cpp:24:20: runtime error: member call on address 0x7ffe95aab8d0 with insufficient space for an object of type 'my_array<int, 0>'
0x7ffe95aab8d0: note: pointer points here
 00 00 00 00  d0 d9 c3 54 48 56 00 00  83 80 10 06 0d 7f 00 00  20 e6 6e 06 0d 7f 00 00  c8 b9 aa 95
              ^ 
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /app/example.cpp:24:20 in 
```

This error is clearly bogus, as an empty type requires a single byte of storage space, and the pointer does not point to the end of the address space. The issue seems to be caused by the `T[0]` member in the `my_array` struct; adding some template parameters to `bug_struct`, or using `std::array` proper instead of the custom type makes the issue go away.

This seems to be a regression: it used to run fine with Clang 15, and it still shows no problems with the current version of GCC.

I'm aware that I shouldn't expect non-standard extensions to always work nicely, but I don't see how the code in the example would have any genuine bug. Looking at the IR, the object size check seems to be folded to an unconditional call to `__ubsan_handle_type_mismatch_v1`, so this might be a codegen/optimizer issue.

**Comment by rymiel** [comment]
It turns out you can reproduce this on 15 too by adding the `-fno-sanitize-recover=object-size` flag: https://godbolt.org/z/cqYza6jh3

**Comment by shafik** [comment]
CC @nickdesaulniers

**Comment by nickdesaulniers** [comment]
My change makes it more likely that code using `__builtin_object_size` is more likely to be inlined.  The decision whether or not to inline isn't itself proof of a compiler bug.

**Comment by BertalanD** [comment]
I did some digging, and 0130b6cb5a8d94511e2bb09ac2f5a613a59f70b4 might be the root cause.

The interesting part is that as `sizeof(my_array<int, 0> == 0)`,  but the IR of the `span<T>` operator seems to assume that the `this` pointer is `dereferenceable(1)`. The actual code that determines the object size for the UBSan check lives here:
https://github.com/llvm/llvm-project/blob/d4879d769095d76ddf9484c50661feac96ce525a/clang/lib/CodeGen/CGClass.cpp#L54-L71

The last return statement changes the returned size to 1 even if the struct is 0-sized. If I trigger the early return above by adding `final` to `my_array`'s declaration, the bogus error disappears. Does anyone know why that `max()` call is there?

**Comment by BertalanD** [comment]
cc @zygoloid maybe

**Comment by zygoloid** [comment]
> I did some digging, and [0130b6c](https://github.com/llvm/llvm-project/commit/0130b6cb5a8d94511e2bb09ac2f5a613a59f70b4) might be the root cause.

Yeah, looks like the heart of the problem here is that the GCC extension permitting arrays of zero length interacts really badly with the rule that's baked into C++ at a fairly fundamental level that all object types have a size of at least 1 (which  that patch is assuming -- and the prior code was too, in some cases).

Here's another way that such types will [fail spectacularly](https://godbolt.org/z/js5xnEG5E):

```
int f(my_array<int, 0> *p, my_array<int, 0> *q) {
  // Crashes, because it computes ((char*)p - (char*)q) / sizeof(my_array<int, 0>).
  return p - q;
}
```

(GCC produces an error for this example.)

Here's [another](https://godbolt.org/z/nxMPE6Pz1). (GCC gives the objects different addresses even though they are zero sized, but can't do so in this case: https://godbolt.org/z/WbWv9jP8e.)

It sounds like we should make a policy decision here of whether we want to fully support types with a size of 0 in C++ everywhere, and if so, fix all the places that are assuming that that doesn't happen. And if not, adding at least an on-by-default warning for that case would seem wise.

**Comment by nickdesaulniers** [comment]
(should this issue be closed now with the addition of 0016f63547f70f969eeb6030d8d7433b05cae934?)

**Comment by nico** [comment]
(Probably not? That's a user-side workaround; the issue in clang remains from what I understand.)
